ext/crates/fp/src/limb.rs

```diff
@@ -11,6 +10,148 @@ pub(crate) struct LimbBitIndexPair {
     pub(crate) bit_index: usize,
 }
 
+/// A struct that defines a range of entries in a slice of limbs. We number limbs starting from some
+/// other limb that might not be in the range. This is because we must be able to index a struct
+/// using indices for another ambient struct. For example, accessing a `Slice` using coordinates
+/// from an ambient `FpVector`.
+#[derive(Debug, Hash, PartialEq, Eq, Clone, Copy)]
+pub struct LimbLength<const P: u32> {
+    /// The index of the first entry relative to some other base limb.
+    pub(crate) start: usize,
+
+    /// The index of the last entry relative to some other base limb.
+    pub(crate) end: usize,
+
+    /// The total number of limbs in the range.
+    ///
+    /// We store this value instead of computing it on the fly because benchmarks tend to show that
+    /// the tradeoff is beneficial in high dimensions (>1000). We might want to only enable this
+    /// when odd-primes is enabled, since computing this number is easier when `p == 2`, so the
+    /// tradeoff is potentially worse.
+    limbs: usize,
+}
+
+impl<const P: u32> LimbLength<P> {
+    pub const fn from_limbs(limbs: usize) -> Self {
+        let logical = entries_per_limb_const::<P>() * limbs;
+        Self {
+            start: 0,
+            end: logical,
+            limbs,
+        }
+    }
+
+    pub const fn from_logical(logical: usize) -> Self {
+        let limbs = number::<P>(logical);
+        Self {
+            start: 0,
+            end: logical,
+            limbs,
+        }
+    }
+
+    /// Returns a `LimbLength` describing a vector starting at entry `start` and ending at entry
+    /// `end`. Most methods assume that `self.start < entries_per_limb_const::<P>()`, so it advised
+    /// to call `apply_shift` before using the return value.
+    pub fn from_start_end(start: usize, end: usize) -> Self {
+        let limb_range = range::<P>(start, end);
+        Self {
+            start,
+            end,
+            limbs: limb_range.end - limb_range.start,
+        }
+    }
+
+    #[inline]
+    pub const fn limbs(&self) -> usize {
+        self.limbs
+    }
+
+    #[inline]
+    pub const fn logical(&self) -> usize {
+        self.end - self.start
+    }
+
+    pub const fn contains(&self, other: &Self) -> bool {
+        self.start <= other.start && self.end >= other.end
+    }
+
+    /// Shift the entire `LimbLength` backwards so that the start of the range belongs to the first
+    /// limb, and return the number of limbs shifted.
+    pub fn apply_shift(&mut self) -> usize {
+        let entries_per = entries_per_limb_const::<P>();
+        let offset = self.start / entries_per;
+        self.start -= offset * entries_per;
+        self.end -= offset * entries_per;
+        offset
+    }
```

Comment:
Would it be cleaner for this to return a new LimbLength and the offset as a tuple? No other functions here mutate the LimbLength

Reply:
Agreed. It technically causes one more unnecessary copy but the compiler should catch that, and in any case it's not performance critical